Source/Scene/ModelOutlineLoader.js

```diff
@@ -0,0 +1,703 @@
+      var outlineData = primitive.extensions.CESIUM_primitive_outline;
+      if (!defined(outlineData)) {
+        return;
+      }
+
+      var vertexNumberingScope = getVertexNumberingScope(model, primitive);
+      if (vertexNumberingScope === undefined) {
+        return;
+      }
+
+      if (vertexNumberingScopes.indexOf(vertexNumberingScope) < 0) {
+        vertexNumberingScopes.push(vertexNumberingScope);
+      }
+
+      // Add the outline to this primitive
+      addOutline(
+        model,
+        meshId,
+        primitiveId,
+        outlineData.indices,
+        vertexNumberingScope
+      );
+    });
+  });
+
+  // Update all relevant bufferViews to include the duplicate vertices that are
+  // needed for outlining.
+  for (var i = 0; i < vertexNumberingScopes.length; ++i) {
+    updateBufferViewsWithNewVertices(
+      model,
+      vertexNumberingScopes[i].bufferViews
+    );
+  }
+
+  // Remove data not referenced by any bufferViews anymore.
+  compactBuffers(model);
+};
+
+ModelOutlineLoader.createTexture = function (model, context) {
+  var cache = context.cache.modelOutliningCache;
+  if (!defined(cache)) {
+    cache = context.cache.modelOutliningCache = {};
+  }
+
+  if (defined(cache.outlineTexture)) {
+    return cache.outlineTexture;
+  }
+
+  var maxSize = Math.min(4096, ContextLimits.maximumTextureSize);
+
+  var size = maxSize;
+  var levelZero = createTexture(size);
+
+  var mipLevels = [];
+
+  while (size > 1) {
+    size >>= 1;
+    mipLevels.push(createTexture(size));
+  }
+
+  var texture = new Texture({
+    context: context,
+    source: {
+      arrayBufferView: levelZero,
+      mipLevels: mipLevels,
+    },
+    width: maxSize,
+    height: 1,
+    pixelFormat: PixelFormat.LUMINANCE,
+    sampler: new Sampler({
+      wrapS: TextureWrap.CLAMP_TO_EDGE,
+      wrapT: TextureWrap.CLAMP_TO_EDGE,
+      minificationFilter: TextureMinificationFilter.LINEAR_MIPMAP_LINEAR,
+      magnificationFilter: TextureMagnificationFilter.LINEAR,
+    }),
+  });
+
+  cache.outlineTexture = texture;
+
+  return texture;
+};
+
+function addOutline(
+  model,
+  meshId,
+  primitiveId,
+  edgeIndicesAccessorId,
+  vertexNumberingScope
+) {
+  var vertexCopies = vertexNumberingScope.vertexCopies;
+  var extraVertices = vertexNumberingScope.extraVertices;
+  var outlineCoordinates = vertexNumberingScope.outlineCoordinates;
+
+  var gltf = model.gltf;
+  var mesh = gltf.meshes[meshId];
+  var primitive = mesh.primitives[primitiveId];
+  var accessors = gltf.accessors;
+  var bufferViews = gltf.bufferViews;
+
+  // Find the number of vertices in this primitive by looking at
+  // the first attribute. Others are required to be the same.
+  var numVertices;
+  for (var semantic in primitive.attributes) {
+    if (primitive.attributes.hasOwnProperty(semantic)) {
+      var attributeId = primitive.attributes[semantic];
+      var accessor = accessors[attributeId];
+      if (defined(accessor)) {
+        numVertices = accessor.count;
```

Comment:
Using the _same_ accessor but different vertices within it is fine. E.g. POSITION is mapped to accessor 0, which is a bufferView with 10 vertices. The first primitive's index buffer, however, only has numbers 0-4, so only the first five vertices in the bufferView are used. Then a second primitive uses the same accessor but its index buffer has indices 5-9, so it uses the last five vertices. No problem. The outliner will add new vertices required by the first primitive to the end of the buffer view, so now the first primitive will use vertices 0-4 plus some new ones starting at 10. Any new vertices for the second primitive will go after those. Not necessarily optimal vertex buffer layout, but nothing breaks.

But there are two related things that can go wrong.

One, that you just made me realize, is that adding new vertices can mean overflowing a uint16 index buffer. I need to handle that.

Two, this code does assume that the accessor `count` reflects the number of vertices in the bufferView. This will always be true except in that dodgy case I mentioned with the "Assumption: A single bufferView contains a single zero-indexed range of vertices..." comment. I _think_ this is an acceptable assumption.

Reply:
I've added the auto-upgrade to 32-bit indices when required, but it's not well tested. I'll do that first thing tomorrow morning (my time zone 😉).